File: econometria espacial/Econometria Espacial.R
# Econometria Espacial em R

# Instalar e carregar pacotes necessários
library(spdep)
library(spatialreg)
library(ggplot2)
library(sf)

# Definir diretório de trabalho
setwd("C:/Users/PC GAMER/Downloads/data")

# Carregar dados
df <- read.csv("data.csv")

# Variáveis
Y <- df$CRIME
X <- df[, c("INC", "HOVAL")]
coords <- cbind(df$X, df$Y)

# Converter para objeto sf para visualização
df_sf <- st_as_sf(df, coords = c("X", "Y"), crs = 4326)

# Criar vizinhança por distância
neighbors <- dnearneigh(coords, d1=0, d2=10)
listw <- nb2listw(neighbors, style="W")

# 🔍 VISUALIZAÇÃO BONITA: mapa de pontos com linhas de vizinhança
# Transformar para objeto spatial (necessário para os gráficos de linhas)
df_sp <- as(df_sf, "Spatial")

# Plotar com ggplot2
plot_neighbors <- function(coords, nb, title = "Mapa de Vizinhança Espacial") {
  segments <- list()
  for (i in seq_along(nb)) {
    for (j in nb[[i]]) {
      segments[[length(segments) + 1]] <- data.frame(
        x = coords[i, 1], y = coords[i, 2],
        xend = coords[j, 1], yend = coords[j, 2]
      )
    }
  }
  seg_df <- do.call(rbind, segments)
  coords_df <- as.data.frame(coords)
  
  ggplot() +
    geom_segment(data = seg_df, aes(x = x, y = y, xend = xend, yend = yend), color = "gray60", alpha = 0.6) +
    geom_point(data = coords_df, aes(x = V1, y = V2), color = "darkred", size = 3) +
    theme_minimal() +
    labs(title = title, x = "Coordenada X", y = "Coordenada Y")
}

# Mostrar gráfico
plot_neighbors(coords, neighbors)

# Estatísticas descritivas
summary(Y)
summary(X)

# OLS
olsreg <- lm(Y ~ INC + HOVAL, data = df)
summary(olsreg)

# Moran's I
moran.test(Y, listw)

# Gráfico de Moran (ajustado)
moran.plot(Y, listw, labels = FALSE, col = "steelblue", pch = 20, xlab = "Variável original", ylab = "Variável espacialmente defasada")

# Testes LM
lm.RStests(olsreg, listw, test = c("LMlag", "LMerr"))

# Modelos espaciais com contiguidade
spatial.lag <- spatialreg::lagsarlm(Y ~ INC + HOVAL, data = df, listw = listw)
summary(spatial.lag)

spatial.error <- spatialreg::errorsarlm(Y ~ INC + HOVAL, data = df, listw = listw)
summary(spatial.error)

# Vizinho baseado em distância (d = 10)
nb_dist <- dnearneigh(coords, d1=0, d2=10)
listw_dist <- nb2listw(nb_dist, style="W")

# Moran com pesos por distância
moran.test(Y, listw_dist)
moran.plot(Y, listw_dist, labels = FALSE, col = "tomato", pch = 16)

# Testes LM
lm.LMtests(olsreg, listw_dist, test = c("LMlag", "LMerr"))

# Modelos espaciais com matriz por distância
spatial.lag_dist <- spatialreg::lagsarlm(Y ~ INC + HOVAL, data = df, listw = listw_dist)
summary(spatial.lag_dist)

spatial.error_dist <- spatialreg::errorsarlm(Y ~ INC + HOVAL, data = df, listw = listw_dist)
summary(spatial.error_dist)

File: econometria espacial/econometria-espacial.jl
using CSV, DataFrames, LinearAlgebra, NearestNeighbors, Plots

# 1. Carregar dados
df = CSV.read("C:/Users/PC GAMER/Downloads/data/data.csv", DataFrame)
coords = hcat(df.X, df.Y)  # Matriz 2 colunas

# 2. Criar estrutura de vizinhança: até distância 10
tree = BallTree(coords', leafsize = 10)  # NearestNeighbors espera linhas = dimensão
dmax = 10.0
neighbors = [findall(i -> i != j && norm(coords[i,:] - coords[j,:]) ≤ dmax, 1:size(coords, 1)) for j in 1:size(coords,1)]

# 3. Construir segmentos (arestas da rede)
segments = Tuple{Float64, Float64, Float64, Float64}[]
for i in 1:length(neighbors)
    for j in neighbors[i]
        push!(segments, (coords[i,1], coords[i,2], coords[j,1], coords[j,2]))
    end
end

# 4. Gerar gráfico
p = plot(; legend=false, title="Rede de Vizinhança Espacial (d ≤ 10)", background_color=:white)
for (x, y, xend, yend) in segments
    plot!([x, xend], [y, yend], lc=:gray60, lw=0.8, alpha=0.6)
end

scatter!(coords[:,1], coords[:,2],
    color=:darkred, markerstrokecolor=:black, markersize=4,
    xlabel="Coordenada X", ylabel="Coordenada Y", framestyle=:box)

# 5. Salvar como PDF
savefig(p, "vizinhanca_espacial.pdf")
